digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/AwsService.scala

```diff
@@ -0,0 +1,301 @@
+            obstacleService.create(newObstacle, AwsUser, link.get)
+          case None =>
+            obstacleService.create(newObstacle, AwsUser, link.get)
+        }
+    }
+  }
+
+  private def updateRoadlink(properties: Map[String, Any], links: Seq[RoadLink]) = {
+    val speedLimit = properties.getOrElse("speedLimit", "").asInstanceOf[String]
+    val pavementClass = properties.getOrElse("pavementClass", "").asInstanceOf[String]
+
+    links.foreach { link =>
+      speedLimit match {
+        case "" =>
+        case _ =>
+          val speedLimitsOnRoadLink = speedLimitService.getExistingAssetByRoadLink(link, false)
+
+          if (speedLimitsOnRoadLink.isEmpty) {
+            val newSpeedLimitAsset = NewLinearAsset(link.linkId, 0, link.length, NumericValue(speedLimit.toInt), properties("sideCode").asInstanceOf[BigInt].intValue(), vvhClient.roadLinkData.createVVHTimeStamp(), None)
+            speedLimitService.createMultiple(Seq(newSpeedLimitAsset), speedLimit.toInt, AwsUser, vvhClient.roadLinkData.createVVHTimeStamp(), (_, _) => Unit)
+          } else
+            speedLimitsOnRoadLink.foreach { sl =>
+              val newSpeedLimitAsset = NewLinearAsset(link.linkId, sl.startMeasure, sl.endMeasure, NumericValue(speedLimit.toInt), properties("sideCode").asInstanceOf[BigInt].intValue(), vvhClient.roadLinkData.createVVHTimeStamp(), None)
+              speedLimitService.update(sl.id, Seq(newSpeedLimitAsset), AwsUser)
+            }
+      }
+
+      pavementClass match {
+        case "" =>
+        case _ =>
+          val pavementClassValue = DynamicValue(DynamicAssetValue(Seq(DynamicProperty("paallysteluokka", "single_choice", required = false, Seq(DynamicPropertyValue(pavementClass.toInt))))))
+
+          val duplicate = pavedRoadService.getPersistedAssetsByLinkIds(PavedRoad.typeId, Seq(link.linkId))
+          duplicate.isEmpty match {
+            case false =>
+              pavedRoadService.update(Seq(duplicate.head.id), pavementClassValue, AwsUser)
+            case true =>
+              val newPavementClassAsset = NewLinearAsset(link.linkId, 0, link.length, pavementClassValue, properties("sideCode").asInstanceOf[BigInt].intValue(), vvhClient.roadLinkData.createVVHTimeStamp(), None)
+              pavedRoadService.create(Seq(newPavementClassAsset), PavedRoad.typeId, AwsUser)
+          }
+      }
+    }
+  }
+
+  def validateAndInsertDataset(dataset: Dataset): Int = {
+    var featuresWithoutIds = 0
+
+    var datasetStatus = 0
+    val assets = dataset.geoJson("features").asInstanceOf[List[Map[String, Any]]]
+    val roadlinks = dataset.roadlinks
+
+    if (assets.length != dataset.roadlinks.length) {
+      datasetStatus = 1
+    }
+
+    awsDao.insertDataset(dataset.datasetId, write(dataset.geoJson), write(dataset.roadlinks), datasetStatus)
+    if (datasetStatus != 1) {
+      var roadlinksCount = 0
+
+      assets.foreach(feature => {
+        val featureRoadlinks = roadlinks(roadlinksCount)
```

Comment:
is the aim here to always treat the nth element of roadlinks and assets at the same time? in that case i think using zipped would be a better option

```
(roadlinks, assets).zipped.foreach((roadlink, asset) => 
	//do stuff
)
```